File: medical-device-regulatory-assistant/backend/tests/integration/api/test_agent_integration.py
"""
Integration tests for CopilotKit Agent Integration
Tests the complete workflow from CopilotKit frontend to LangGraph backend
"""

import pytest
import asyncio
import json
from unittest.mock import AsyncMock, patch, MagicMock
from fastapi.testclient import TestClient

from main import app
from agents.regulatory_agent import RegulatoryAgent
from agents.regulatory_agent_state import AgentTaskType, AgentStatus
from services.session_manager import SessionManager
from api.agent_integration import get_session_manager


@pytest.fixture
def client():
    """Test client for FastAPI app"""
    with TestClient(app) as client:
        yield client


@pytest.fixture
def mock_agent():
    """Mock regulatory agent"""
    agent = AsyncMock(spec=RegulatoryAgent)
    
    # Mock start_session
    agent.start_session.return_value = {
        "session_id": "test-session-123",
        "status": "initialized",
        "context": {"device_description": "Test device"}
    }
    
    # Mock execute_task
    agent.execute_task.return_value = {
        "session_id": "test-session-123",
        "task_type": "predicate_search",
        "status": "completed",
        "result": {
            "predicates": [
                {
                    "k_number": "K123456",
                    "device_name": "Test Predicate Device",
                    "confidence_score": 0.85
                }
            ]
        },
        "confidence": 0.85,
        "execution_time_ms": 2500
    }
    
    # Mock get_session_state
    agent.get_session_state.return_value = {
        "session_id": "test-session-123",
        "status": "completed",
        "completed_tasks": ["predicate_search"],
        "project_id": "test-project"
    }
    
    return agent


@pytest.fixture
def mock_session_manager():
    """Mock session manager"""
    manager = AsyncMock(spec=SessionManager)
    
    manager.get_session.return_value = None  # No existing session
    manager.store_session.return_value = None
    manager.get_active_session_count.return_value = 1
    
    return manager


class TestAgentIntegration:
    """Test agent integration endpoints"""
    
    def test_execute_agent_task_success(self, client, mock_agent):
        """Test successful agent task execution"""
        
        with patch('backend.api.agent_integration.session_manager') as mock_sm:
            mock_sm.get_session.return_value = None
            mock_sm.store_session.return_value = None
            
            with patch('backend.api.agent_integration.RegulatoryAgent', return_value=mock_agent):
                
                request_data = {
                    "task_type": "predicate_search",
                    "project_id": "test-project",
                    "device_description": "Cardiac monitoring device",
                    "intended_use": "For continuous heart rhythm monitoring",
                    "parameters": {"product_code": "DQK"}
                }
                
                response = client.post(
                    "/api/agent/execute",
                    json=request_data,
                    headers={"Authorization": "Bearer test-token"}
                )
                
                assert response.status_code == 200
                data = response.json()
                
                assert data["session_id"] == "test-session-123"
                assert data["task_type"] == "predicate_search"
                assert data["status"] == "completed"
                assert data["confidence"] == 0.85
                assert "result" in data
    
    def test_execute_agent_task_invalid_task_type(self, client):
        """Test agent task execution with invalid task type"""
        
        request_data = {
            "task_type": "invalid_task",
            "project_id": "test-project",
            "device_description": "Test device",
            "intended_use": "Test use"
        }
        
        response = client.post(
            "/api/agent/execute",
            json=request_data,
            headers={"Authorization": "Bearer test-token"}
        )
        
        assert response.status_code == 400
        assert "Invalid task type" in response.json()["detail"]
    
    def test_get_session_status(self, client, mock_agent):
        """Test getting session status"""
        
        with patch('backend.api.agent_integration.session_manager') as mock_sm:
            mock_sm.get_session.return_value = mock_agent
            
            response = client.get(
                "/api/agent/session/test-session-123/status",
                headers={"Authorization": "Bearer test-token"}
            )
            
            assert response.status_code == 200
            data = response.json()
            
            assert data["session_id"] == "test-session-123"
            assert data["status"] == "completed"
            assert "completed_tasks" in data
    
    def test_get_session_status_not_found(self, client):
        """Test getting status for non-existent session"""
        
        with patch('backend.api.agent_integration.session_manager') as mock_sm:
            mock_sm.get_session.return_value = None
            
            response = client.get(
                "/api/agent/session/nonexistent/status",
                headers={"Authorization": "Bearer test-token"}
            )
            
            assert response.status_code == 404
    
    def test_cancel_session_task(self, client, mock_agent):
        """Test cancelling a session task"""
        
        with patch('backend.api.agent_integration.session_manager') as mock_sm:
            mock_sm.get_session.return_value = mock_agent
            
            # Mock session state as processing
            mock_agent.get_session_state.return_value = {
                "session_id": "test-session-123",
                "status": "processing",
                "project_id": "test-project"
            }
            
            request_data = {
                "session_id": "test-session-123",
                "reason": "User requested cancellation"
            }
            
            response = client.post(
                "/api/agent/session/test-session-123/cancel",
                json=request_data,
                headers={"Authorization": "Bearer test-token"}
            )
            
            assert response.status_code == 200
            data = response.json()
            
            assert data["session_id"] == "test-session-123"
            assert data["status"] == "cancelled"
    
    def test_health_check(self, client):
        """Test agent integration health check"""
        
        with patch('backend.api.agent_integration.session_manager') as mock_sm:
            mock_sm.get_active_session_count.return_value = 5
            
            response = client.get("/api/agent/health")
            
            assert response.status_code == 200
            data = response.json()
            
            assert data["status"] == "healthy"
            assert data["active_sessions"] == 5
            assert "timestamp" in data


class TestCopilotKitIntegration:
    """Test CopilotKit integration with backend"""
    
    def test_copilotkit_predicate_search(self, client):
        """Test CopilotKit predicate search action"""
        
        # Mock the backend API call
        with patch('requests.post') as mock_post:
            mock_response = MagicMock()
            mock_response.ok = True
            mock_response.json.return_value = {
                "session_id": "test-session",
                "task_type": "predicate_search",
                "status": "completed",
                "result": {
                    "predicates": [
                        {
                            "k_number": "K123456",
                            "device_name": "Test Device",
                            "confidence_score": 0.85
                        }
                    ]
                },
                "confidence": 0.85
            }
            mock_post.return_value = mock_response
            
            # Test the CopilotKit endpoint
            response = client.post(
                "/api/copilotkit",
                json={
                    "action": "predicate_search",
                    "parameters": {
                        "deviceDescription": "Cardiac monitoring device",
                        "intendedUse": "Heart rhythm monitoring"
                    }
                }
            )
            
            # Note: This test would need to be adapted based on 
            # CopilotKit's actual request/response format
            assert response.status_code in [200, 404]  # 404 if route not properly set up


class TestRealTimeUpdates:
    """Test real-time status updates via Server-Sent Events"""
    
    def test_session_stream_events(self, client, mock_agent):
        """Test SSE stream for session updates"""
        
        with patch('backend.api.agent_integration.session_manager') as mock_sm:
            mock_sm.get_session.return_value = mock_agent
            
            # Mock session state changes
            mock_agent.get_session_state.side_effect = [
                {
                    "session_id": "test-session",
                    "status": "processing",
                    "updated_at": "2024-01-01T10:00:00"
                },
                {
                    "session_id": "test-session", 
                    "status": "completed",
                    "updated_at": "2024-01-01T10:01:00"
                }
            ]
            
            # Test SSE endpoint
            response = client.get(
                "/api/agent/session/test-session/stream",
                headers={"Authorization": "Bearer test-token"}
            )
            
            # SSE endpoints typically return 200 and stream data
            assert response.status_code == 200


class TestErrorHandling:
    """Test error handling in agent integration"""
    
    def test_agent_execution_error(self, client):
        """Test handling of agent execution errors"""
        
        with patch('backend.api.agent_integration.RegulatoryAgent') as mock_agent_class:
            mock_agent = AsyncMock()
            mock_agent.start_session.side_effect = Exception("Agent initialization failed")
            mock_agent_class.return_value = mock_agent
            
            request_data = {
                "task_type": "predicate_search",
                "project_id": "test-project",
                "device_description": "Test device",
                "intended_use": "Test use"
            }
            
            response = client.post(
                "/api/agent/execute",
                json=request_data,
                headers={"Authorization": "Bearer test-token"}
            )
            
            assert response.status_code == 500
            assert "Agent task execution failed" in response.json()["detail"]
    
    def test_unauthorized_access(self, client):
        """Test unauthorized access to agent endpoints"""
        
        request_data = {
            "task_type": "predicate_search",
            "project_id": "test-project",
            "device_description": "Test device",
            "intended_use": "Test use"
        }
        
        # No authorization header
        response = client.post("/api/agent/execute", json=request_data)
        
        # Should still work with default test user in MVP
        # In production, this would return 401
        assert response.status_code in [200, 401, 500]


if __name__ == "__main__":
    pytest.main([__file__, "-v"])

File: medical-device-regulatory-assistant/scripts/test-agent-integration.js
/**
 * Integration test script for CopilotKit Agent Integration
 * Tests the complete workflow from frontend to backend
 */

const fetch = require('node-fetch');

const BACKEND_URL = process.env.BACKEND_URL || 'http://localhost:8000';
const FRONTEND_URL = process.env.FRONTEND_URL || 'http://localhost:3000';

async function testAgentIntegration() {
  console.log('🧪 Testing CopilotKit Agent Integration...\n');

  // Test 1: Backend Health Check
  console.log('1. Testing backend health check...');
  try {
    const response = await fetch(`${BACKEND_URL}/api/agent/health`);
    if (response.ok) {
      const data = await response.json();
      console.log('✅ Backend health check passed');
      console.log(`   Status: ${data.status}`);
      console.log(`   Active sessions: ${data.active_sessions}`);
    } else {
      console.log('❌ Backend health check failed');
      console.log(`   Status: ${response.status} ${response.statusText}`);
    }
  } catch (error) {
    console.log('❌ Backend health check failed');
    console.log(`   Error: ${error.message}`);
  }

  console.log();

  // Test 2: Agent Task Execution
  console.log('2. Testing agent task execution...');
  try {
    const taskRequest = {
      task_type: 'predicate_search',
      project_id: 'integration-test-project',
      device_description: 'Cardiac monitoring device for integration testing',
      intended_use:
        'For continuous monitoring of cardiac rhythm in ambulatory patients',
      device_type: 'Class II Medical Device',
      parameters: {
        product_code: 'DQK',
      },
    };

    const response = await fetch(`${BACKEND_URL}/api/agent/execute`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: 'Bearer test-token',
      },
      body: JSON.stringify(taskRequest),
    });

    if (response.ok) {
      const data = await response.json();
      console.log('✅ Agent task execution passed');
      console.log(`   Session ID: ${data.session_id}`);
      console.log(`   Task Type: ${data.task_type}`);
      console.log(`   Status: ${data.status}`);
      console.log(`   Confidence: ${data.confidence || 'N/A'}`);

      // Test 3: Session Status Check
      console.log('\n3. Testing session status check...');
      const statusResponse = await fetch(
        `${BACKEND_URL}/api/agent/session/${data.session_id}/status`,
        {
          headers: {
            Authorization: 'Bearer test-token',
          },
        }
      );

      if (statusResponse.ok) {
        const statusData = await statusResponse.json();
        console.log('✅ Session status check passed');
        console.log(`   Session Status: ${statusData.status}`);
        console.log(
          `   Completed Tasks: ${statusData.completed_tasks?.length || 0}`
        );
      } else {
        console.log('❌ Session status check failed');
        console.log(
          `   Status: ${statusResponse.status} ${statusResponse.statusText}`
        );
      }
    } else {
      console.log('❌ Agent task execution failed');
      console.log(`   Status: ${response.status} ${response.statusText}`);
      const errorData = await response.text();
      console.log(`   Error: ${errorData}`);
    }
  } catch (error) {
    console.log('❌ Agent task execution failed');
    console.log(`   Error: ${error.message}`);
  }

  console.log();

  // Test 4: CopilotKit API Endpoint
  console.log('4. Testing CopilotKit API endpoint...');
  try {
    const response = await fetch(`${FRONTEND_URL}/api/copilotkit`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        // This would be the actual CopilotKit request format
        // For now, just test that the endpoint exists
        test: true,
      }),
    });

    // CopilotKit endpoints typically return different status codes
    // We're just checking if the endpoint is accessible
    console.log(
      `✅ CopilotKit endpoint accessible (Status: ${response.status})`
    );
  } catch (error) {
    if (error.code === 'ECONNREFUSED') {
      console.log('⚠️  Frontend server not running (this is expected in CI)');
    } else {
      console.log('❌ CopilotKit endpoint test failed');
      console.log(`   Error: ${error.message}`);
    }
  }

  console.log();

  // Test 5: Tool Registry Health
  console.log('5. Testing tool registry health...');
  try {
    const response = await fetch(`${BACKEND_URL}/api/health`);
    if (response.ok) {
      const data = await response.json();
      console.log('✅ Tool registry health check passed');
      console.log(`   Service: ${data.service}`);
      console.log(`   Version: ${data.version}`);
    } else {
      console.log('❌ Tool registry health check failed');
    }
  } catch (error) {
    console.log('❌ Tool registry health check failed');
    console.log(`   Error: ${error.message}`);
  }

  console.log('\n🏁 Integration test completed!\n');
}

// Run the test
if (require.main === module) {
  testAgentIntegration().catch(console.error);
}

module.exports = { testAgentIntegration };
